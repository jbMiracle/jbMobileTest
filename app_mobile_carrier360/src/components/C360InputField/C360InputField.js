import React, { Component } from 'react';
import { Animated, View, Vibration } from 'react-native';
import { Text, Input, connectStyle } from 'native-base';
import PropTypes from 'prop-types';
import { omit, isFunction } from 'lodash';
import * as Animatable from 'react-native-animatable';
import { StyledC360Icon } from '../C360Icon/C360Icon';
import C360Colors from '../../themes/360Colors';
import styleDef from './C360InputField.style';

class C360InputField extends Component {
  static propTypes = {
    placeHolderText: PropTypes.string.isRequired,
    onChangeText: PropTypes.func,
    value: PropTypes.string,
    leftIcon: PropTypes.string,
    error: PropTypes.bool,
    errorVibrate: PropTypes.bool,
    dark: PropTypes.bool,
  };

  static defaultProps = {
    onChangeText: undefined,
    value: null,
    leftIcon: null,
    error: false,
    errorVibrate: false,
    dark: false,
  };

  state = {
    showTitle: false,
    topAnimationValue: new Animated.Value(2),
  };

  componentWillMount() {
    const { value } = this.props;
    if (value) {
      this.setState({ showTitle: true, text: value });
      this.showTitle();
    }
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.error && !this.props.error) {
      this.animationViewRef.shake();

      if (this.props.errorVibrate) {
        Vibration.vibrate();
      }
    }
    if (nextProps.value && nextProps.value !== this.state.text) {
      this.setState(state => ({ ...state, text: nextProps.value }));
    }
  }

  onFocus = () => {
    const { onFocus } = this.inputProps;
    this.setState(() => ({ showTitle: true }));
    if (isFunction(onFocus)) {
      onFocus();
    }
  };

  onBlur = () => {
    const { onBlur } = this.inputProps;
    this.setState(state => ({ showTitle: !!state.text }));
    if (isFunction(onBlur)) {
      onBlur();
    }
  };

  onChangeText = (text) => {
    const { onChangeText } = this.inputProps;
    this.setState(state => ({ ...state, text }));
    if (isFunction(onChangeText)) {
      onChangeText(text);
    }
    if (isFunction(this.props.onChangeText)) {
      this.props.onChangeText(text);
    }
  };

  // This ref function breaks styleguide, but this paradigm is required for use of Animated Library.
  // See https://github.com/oblador/react-native-animatable#imperative-usage
  getAnimationViewRef = (ref) => this.animationViewRef = ref; // eslint-disable-line

  animateTitle(toValue) {
    Animated.timing(this.state.topAnimationValue, { toValue, duration: 150 }).start();
  }

  hideTitle() {
    this.animateTitle(2);
  }

  showTitle() {
    this.animateTitle(30);
  }

  renderInput = () => (
    <Input
      key='input'
      value={ this.state.text }
      selectionColor={ this.props.dark ? C360Colors.White : C360Colors.Grey600 } /* Will override if passed in the props */
      { ...this.inputProps }
      onFocus={ this.onFocus }
      onBlur={ this.onBlur }
      onChangeText={ this.onChangeText }
    />
  );

  renderBody(leftIconProps, rightIconProps) {
    const { dark, placeHolderText } = this.props;
    const { showTitle, topAnimationValue } = this.state;
    if (showTitle) {
      this.showTitle();
    } else {
      this.hideTitle();
    }
    const leftSpaceStyle = leftIconProps ? styleDef.leftMargin : styleDef.leftAlign;
    const rightSpaceStyle = rightIconProps ? styleDef.rightMargin : styleDef.rightAlign;
    return (
      <View
        key='view'
        style={ [styleDef.body, leftSpaceStyle, rightSpaceStyle] }
      >
        <Animated.View style={ [styleDef.animatedView, { bottom: topAnimationValue }] }>
          <Text
            key='text'
            style={ showTitle
              ? [styleDef.title, !dark && styleDef.blueFont]
              : [styleDef.placeHolder, !dark && styleDef.greyFont]
            }
          >
            {showTitle ? placeHolderText.toUpperCase() : placeHolderText}
          </Text>
        </Animated.View>
        {this.renderInput()}
      </View>
    );
  }

  renderChildren() {
    const { leftIcon, error } = this.props;
    const leftIconProps = leftIcon ? ({ name: leftIcon }) : null;
    const rightIconProps = error ? styleDef.errorIcon : null;
    const parentPropTypes = this.constructor.propTypes;
    const keysToBeOmitted = [...Object.keys(parentPropTypes), 'style'];
    /* This is done to filter TextInput props. Eg: secureTextEntry, maxLength, etc., */
    this.inputProps = omit(this.props, keysToBeOmitted);
    return [
      leftIconProps ? this.renderSideIcon(leftIconProps) : null,
      this.renderBody(leftIconProps, rightIconProps),
      rightIconProps ? this.renderSideIcon(rightIconProps) : null,
    ];
  }

  renderSideIcon = props => (
    <StyledC360Icon key={ props.name } { ...props } />
  );

  render() {
    return (
      <Animatable.View
        ref={ this.getAnimationViewRef }
      >
        <View
          ref={ (component) => {
            this.root = component;
          } }
          { ...this.props }
        >
          {this.renderChildren()}
        </View>
      </Animatable.View>
    );
  }
}

export const StyledInputField = connectStyle('C360.InputField', {})(C360InputField);
export const AnimatableStyledInputField = Animatable.createAnimatableComponent(StyledInputField);
export { AnimatableStyledInputField as C360InputField };
