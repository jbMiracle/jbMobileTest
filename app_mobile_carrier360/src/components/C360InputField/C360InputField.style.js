import variable from '../../themes/defaultTheme/variables/platform';
import C360Colors from '../../themes/360Colors';

export default {
  errorIcon: {
    name: 'jbh-Circle_Warning',
    style: { color: C360Colors.Red700 },
  },
  animatedView: {
    position: 'absolute',
  },
  title: {
    fontSize: 10,
    fontFamily: variable.headerFontFamily,
  },
  placeHolder: {
    fontSize: 16,
    fontFamily: variable.fontFamily,
  },
  body: {
    flex: 1,
  },
  leftMargin: {
    marginLeft: 12,
  },
  rightMargin: {
    marginRight: 12,
  },
  leftAlign: {
    marginLeft: 0,
  },
  rightAlign: {
    marginRight: 0,
  },
  blueFont: {
    color: C360Colors.Blue500,
  },
  greyFont: {
    color: C360Colors.Grey600,
  },
  whiteFont: {
    color: C360Colors.White,
  },
};
