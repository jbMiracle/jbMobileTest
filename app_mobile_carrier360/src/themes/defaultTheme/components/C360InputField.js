import { Platform } from 'react-native';
import variable from './../variables/platform';
import C360Colors from '../../../themes/360Colors';

export default (variables = variable) => ({
  '.dark': {
    'NativeBase.Input': {
      color: C360Colors.White,
    },
    'NativeBase.Text': {
      color: C360Colors.White,
    },
    'NativeBase.IconNB': {
      color: C360Colors.White,
    },
    borderColor: C360Colors.White,
  },
  'NativeBase.Input': {
    color: C360Colors.Black,
    fontSize: variables.DefaultFontSize,
  },
  'NativeBase.Text': {
    color: C360Colors.Grey600,
    fontSize: variables.DefaultFontSize,
  },
  'NativeBase.IconNB': {
    color: C360Colors.Grey600,
    alignSelf: 'center',
    fontSize: variables.DefaultFontSize,
  },
  '.error': {
    borderColor: C360Colors.Red700,
  },
  borderColor: C360Colors.Grey600,
  borderBottomWidth: 2,
  flexDirection: 'row',
  width: '100%',
  paddingBottom: 8,
  paddingTop: 12,
  marginTop: Platform.OS === 'ios' ? 20 : 12,
});
