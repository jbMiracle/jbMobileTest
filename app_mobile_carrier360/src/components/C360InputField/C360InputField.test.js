import React from 'react';
import TestRenderer from 'react-test-renderer';
import { C360InputField } from './C360InputField';

jest.useFakeTimers();

describe('Can render Input Field', () => {
  it('render without icon', () => {
    const props = {
      id: 'inputField',
      placeHolderText: 'Username',
      value: null,
      leftIconName: null,
      error: false,
      dark: false,
    };
    const rendered = TestRenderer.create(<C360InputField { ...props } />);
    expect(rendered.toJSON()).toMatchSnapshot();
  });
  it('render with icon', () => {
    const props = {
      id: 'inputField',
      placeHolderText: 'Username',
      value: null,
      leftIconName: 'jbh-User_Profile',
      error: false,
      dark: false,
    };
    const rendered = TestRenderer.create(<C360InputField { ...props } />);
    expect(rendered.toJSON()).toMatchSnapshot();
  });
  it('render with value and icon', () => {
    const props = {
      id: 'inputField',
      placeHolderText: 'Username',
      value: 'Demo',
      leftIconName: 'jbh-User_Profile',
      error: false,
      dark: false,
    };
    const rendered = TestRenderer.create(<C360InputField { ...props } />);
    expect(rendered.toJSON()).toMatchSnapshot();
  });
  it('render dark theme', () => {
    const props = {
      id: 'inputField',
      placeHolderText: 'Username',
      value: 'Demo',
      leftIconName: 'jbh-User_Profile',
      error: false,
      dark: true,
    };
    const rendered = TestRenderer.create(<C360InputField { ...props } />);
    expect(rendered.toJSON()).toMatchSnapshot();
  });
  it('render with error', () => {
    const props = {
      id: 'inputField',
      placeHolderText: 'Username',
      value: 'Demo',
      leftIconName: 'jbh-User_Profile',
      error: true,
      dark: false,
    };
    const rendered = TestRenderer.create(<C360InputField { ...props } />);
    expect(rendered.toJSON()).toMatchSnapshot();
  });
});
