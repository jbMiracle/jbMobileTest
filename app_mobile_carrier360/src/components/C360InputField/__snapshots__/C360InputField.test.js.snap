// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Can render Input Field render dark theme 1`] = `
<View
  collapsable={undefined}
  style={Object {}}
>
  <View
    collapsable={undefined}
    dark={true}
    error={false}
    errorVibrate={false}
    id="inputField"
    leftIcon={null}
    leftIconName="jbh-User_Profile"
    onChangeText={undefined}
    placeHolderText="Username"
    style={
      Array [
        Object {},
        Object {},
      ]
    }
    value="Demo"
    virtual={undefined}
  >
    <View
      style={
        Array [
          Object {
            "flex": 1,
          },
          Object {
            "marginLeft": 0,
          },
          Object {
            "marginRight": 0,
          },
        ]
      }
    >
      <View
        collapsable={undefined}
        style={
          Object {
            "bottom": 2,
            "position": "absolute",
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "color": "#000",
                "fontFamily": "System",
                "fontSize": 16,
              },
              Object {
                "fontFamily": "ProximaNova-Bold",
                "fontSize": 10,
              },
              false,
            ]
          }
          uppercase={false}
          virtual={undefined}
        >
          USERNAME
        </Text>
      </View>
      <TextInput
        allowFontScaling={true}
        collapsable={undefined}
        editable={true}
        id="inputField"
        leftIconName="jbh-User_Profile"
        onBlur={[Function]}
        onChangeText={[Function]}
        onFocus={[Function]}
        placeholderTextColor="#575757"
        selectionColor="#FFFFFF"
        style={
          Object {
            "color": "#000",
            "flex": 1,
            "fontSize": 17,
            "height": 50,
            "lineHeight": 24,
            "paddingLeft": 5,
            "paddingRight": 5,
          }
        }
        underlineColorAndroid="rgba(0,0,0,0)"
        value="Demo"
        virtual={undefined}
      />
    </View>
  </View>
</View>
`;

exports[`Can render Input Field render with error 1`] = `
<View
  collapsable={undefined}
  style={Object {}}
>
  <View
    collapsable={undefined}
    dark={false}
    error={true}
    errorVibrate={false}
    id="inputField"
    leftIcon={null}
    leftIconName="jbh-User_Profile"
    onChangeText={undefined}
    placeHolderText="Username"
    style={
      Array [
        Object {},
        Object {},
      ]
    }
    value="Demo"
    virtual={undefined}
  >
    <View
      style={
        Array [
          Object {
            "flex": 1,
          },
          Object {
            "marginLeft": 0,
          },
          Object {
            "marginRight": 12,
          },
        ]
      }
    >
      <View
        collapsable={undefined}
        style={
          Object {
            "bottom": 2,
            "position": "absolute",
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "color": "#000",
                "fontFamily": "System",
                "fontSize": 16,
              },
              Object {
                "fontFamily": "ProximaNova-Bold",
                "fontSize": 10,
              },
              Object {
                "color": "#007DBA",
              },
            ]
          }
          uppercase={false}
          virtual={undefined}
        >
          USERNAME
        </Text>
      </View>
      <TextInput
        allowFontScaling={true}
        collapsable={undefined}
        editable={true}
        id="inputField"
        leftIconName="jbh-User_Profile"
        onBlur={[Function]}
        onChangeText={[Function]}
        onFocus={[Function]}
        placeholderTextColor="#575757"
        selectionColor="#939393"
        style={
          Object {
            "color": "#000",
            "flex": 1,
            "fontSize": 17,
            "height": 50,
            "lineHeight": 24,
            "paddingLeft": 5,
            "paddingRight": 5,
          }
        }
        underlineColorAndroid="rgba(0,0,0,0)"
        value="Demo"
        virtual={undefined}
      />
    </View>
    <Text
      accessible={true}
      allowFontScaling={false}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": undefined,
            "fontSize": 12,
          },
          Array [
            Object {
              "color": "#000",
              "fontSize": 30,
            },
            Object {
              "color": "#993A29",
            },
          ],
          Object {
            "fontFamily": "icomoon",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
        ]
      }
      virtual={undefined}
    >
      î¦œ
    </Text>
  </View>
</View>
`;

exports[`Can render Input Field render with icon 1`] = `
<View
  collapsable={undefined}
  style={Object {}}
>
  <View
    collapsable={undefined}
    dark={false}
    error={false}
    errorVibrate={false}
    id="inputField"
    leftIcon={null}
    leftIconName="jbh-User_Profile"
    onChangeText={undefined}
    placeHolderText="Username"
    style={
      Array [
        Object {},
        Object {},
      ]
    }
    value={null}
    virtual={undefined}
  >
    <View
      style={
        Array [
          Object {
            "flex": 1,
          },
          Object {
            "marginLeft": 0,
          },
          Object {
            "marginRight": 0,
          },
        ]
      }
    >
      <View
        collapsable={undefined}
        style={
          Object {
            "bottom": 2,
            "position": "absolute",
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "color": "#000",
                "fontFamily": "System",
                "fontSize": 16,
              },
              Object {
                "fontFamily": "ProximaNova-Regular",
                "fontSize": 16,
              },
              Object {
                "color": "#939393",
              },
            ]
          }
          uppercase={false}
          virtual={undefined}
        >
          Username
        </Text>
      </View>
      <TextInput
        allowFontScaling={true}
        collapsable={undefined}
        editable={true}
        id="inputField"
        leftIconName="jbh-User_Profile"
        onBlur={[Function]}
        onChangeText={[Function]}
        onFocus={[Function]}
        placeholderTextColor="#575757"
        selectionColor="#939393"
        style={
          Object {
            "color": "#000",
            "flex": 1,
            "fontSize": 17,
            "height": 50,
            "lineHeight": 24,
            "paddingLeft": 5,
            "paddingRight": 5,
          }
        }
        underlineColorAndroid="rgba(0,0,0,0)"
        value={undefined}
        virtual={undefined}
      />
    </View>
  </View>
</View>
`;

exports[`Can render Input Field render with value and icon 1`] = `
<View
  collapsable={undefined}
  style={Object {}}
>
  <View
    collapsable={undefined}
    dark={false}
    error={false}
    errorVibrate={false}
    id="inputField"
    leftIcon={null}
    leftIconName="jbh-User_Profile"
    onChangeText={undefined}
    placeHolderText="Username"
    style={
      Array [
        Object {},
        Object {},
      ]
    }
    value="Demo"
    virtual={undefined}
  >
    <View
      style={
        Array [
          Object {
            "flex": 1,
          },
          Object {
            "marginLeft": 0,
          },
          Object {
            "marginRight": 0,
          },
        ]
      }
    >
      <View
        collapsable={undefined}
        style={
          Object {
            "bottom": 2,
            "position": "absolute",
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "color": "#000",
                "fontFamily": "System",
                "fontSize": 16,
              },
              Object {
                "fontFamily": "ProximaNova-Bold",
                "fontSize": 10,
              },
              Object {
                "color": "#007DBA",
              },
            ]
          }
          uppercase={false}
          virtual={undefined}
        >
          USERNAME
        </Text>
      </View>
      <TextInput
        allowFontScaling={true}
        collapsable={undefined}
        editable={true}
        id="inputField"
        leftIconName="jbh-User_Profile"
        onBlur={[Function]}
        onChangeText={[Function]}
        onFocus={[Function]}
        placeholderTextColor="#575757"
        selectionColor="#939393"
        style={
          Object {
            "color": "#000",
            "flex": 1,
            "fontSize": 17,
            "height": 50,
            "lineHeight": 24,
            "paddingLeft": 5,
            "paddingRight": 5,
          }
        }
        underlineColorAndroid="rgba(0,0,0,0)"
        value="Demo"
        virtual={undefined}
      />
    </View>
  </View>
</View>
`;

exports[`Can render Input Field render without icon 1`] = `
<View
  collapsable={undefined}
  style={Object {}}
>
  <View
    collapsable={undefined}
    dark={false}
    error={false}
    errorVibrate={false}
    id="inputField"
    leftIcon={null}
    leftIconName={null}
    onChangeText={undefined}
    placeHolderText="Username"
    style={
      Array [
        Object {},
        Object {},
      ]
    }
    value={null}
    virtual={undefined}
  >
    <View
      style={
        Array [
          Object {
            "flex": 1,
          },
          Object {
            "marginLeft": 0,
          },
          Object {
            "marginRight": 0,
          },
        ]
      }
    >
      <View
        collapsable={undefined}
        style={
          Object {
            "bottom": 2,
            "position": "absolute",
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "color": "#000",
                "fontFamily": "System",
                "fontSize": 16,
              },
              Object {
                "fontFamily": "ProximaNova-Regular",
                "fontSize": 16,
              },
              Object {
                "color": "#939393",
              },
            ]
          }
          uppercase={false}
          virtual={undefined}
        >
          Username
        </Text>
      </View>
      <TextInput
        allowFontScaling={true}
        collapsable={undefined}
        editable={true}
        id="inputField"
        leftIconName={null}
        onBlur={[Function]}
        onChangeText={[Function]}
        onFocus={[Function]}
        placeholderTextColor="#575757"
        selectionColor="#939393"
        style={
          Object {
            "color": "#000",
            "flex": 1,
            "fontSize": 17,
            "height": 50,
            "lineHeight": 24,
            "paddingLeft": 5,
            "paddingRight": 5,
          }
        }
        underlineColorAndroid="rgba(0,0,0,0)"
        value={undefined}
        virtual={undefined}
      />
    </View>
  </View>
</View>
`;
